pypiserver/_app.py

```diff
@@ -357,25 +406,56 @@ def list_packages():
     return template(tmpl, links=links)
 
 
-@app.route("/packages/:filename#.*#")
+@app.route(f"/packages/<pkg_version:re:(.*-.*)-.*-.*-.*\.whl\.metadata>")
+@auth("download")
+def serve_metadata(pkg_version):
+    pkg, version, _ = pkg_version.split("-", 2)
+    urs = urlsplit(config.fallback_url.rstrip("/"))
+    json_url = f"{urs.scheme}://{urs.hostname}/pypi/{pkg}/json"
+    response = urllib.request.urlopen(json_url)
+    content = json.loads(response.read().decode("utf-8"))
+    metadata_url = [
+        i["url"]
+        for i in content["releases"][version]
+        if i["packagetype"] == "bdist_wheel"
+    ][0]
+    response = urllib.request.urlopen(metadata_url + ".metadata")
+    content = response.read().decode("utf-8")
+    response = HTTPResponse(body=content, status=200)
+    return response
+
+
+@app.route("/packages/<filename:re:.*(whl|tar.gz|zip)>")
 @auth("download")
 def server_static(filename):
+    response = HTTPError(404, f"Not Found ({filename} does not exist)\n\n")
     entries = config.backend.get_all_packages()
     for x in entries:
         f = x.relfn_unix
         if f == filename:
+            log.debug(f"Package {filename} found localy, serving")
             response = static_file(
                 filename,
                 root=x.root,
                 mimetype=mimetypes.guess_type(filename)[0],
             )
-            if config.cache_control:
-                response.set_header(
-                    "Cache-Control", f"public, max-age={config.cache_control}"
-                )
-            return response
+    if response.status_code == 404 and config.pull_through:
+        # Fetche the package from pypi.org
+        # and save it the local filesystem
+        log.debug(f"Package {filename} not found localy, pulling it")
+        fetch_package(filename, str(config.roots[0]))
+        response = static_file(
+            filename,
+            root=str(config.roots[0]),
+            mimetype=mimetypes.guess_type(filename)[0],
+        )
+
+    if config.cache_control:
+        response.set_header(
+            "Cache-Control", f"public, max-age={config.cache_control}"
+        )
```

Comment:
[curious] This is really nice. So in this case, we will download the `pull-through` version from a fallback URL and store it as if it was a local package. That's neat 👍  Is that correct?

Reply:
Yes. That is correct. 
Let's say you start with an empty local directory. If you point pip to your local pypiserver and try to install a package foo.
It will not be found, and then it will be fetched with metadata for pypi.org.
If you will try a second time (if there is no local pip cache) the package will be served from pypiserver without first going to get the package from pypi.org.
(Note, you can replace pypi.org with any upstream).